Refactor sharding data pipe into a seperate file

wenleix opened the pull request:
Move `ShardingFilterIterDataPipe` into a dedicated file. 

Also, propose to have a dedicated parent class (`_ShardingIterDataPipe`) for sharding data pipe, as this seems more like a "system/engine-level" datapipe that gives strong hints to RS on how to execute, and needs first-class citizen treatment in RS (compared with other "user-level" datapipe that are mostly composable `Callable[[Iterable], Iterable]`.  So we don't need to based on whether `is_shardable` and `apply_sharding` are presented in DataPipe in `graph_settings.py`. But open to other discussions. 



Open question: Should 
[ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17) also be considered as a `_ShardingIterDataPipe`? (e.g. this sharding is executed by replicating (the metadata), while `ShardingRoundRobinDispatcherIterDataPipe` hints too expensive to replicate so requires round robin data exchange/dispatch). 

Differential Revision: D43014692

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -23,67 +22,6 @@ class SHARDING_PRIORITIES(IntEnum):
     MULTIPROCESSING = 3
 
 
-@functional_datapipe('sharding_filter')
```
Let's move `SHARDING_PRIORITIES` as well. And, if we do so, we need to land the changes for torchdata.

wenleix replied:
> And, if we do so, we need to land the changes for torchdata.

What kind of changes? -- Change the import right? . What's the procedure of doing that usually? ~

ejguan replied:
> Change the import right?

Yeah

> What's the procedure of doing that usually? 

Open a Diff that contains changes for both pytorch and torchdata. Then, land this diff directly. The CI for TorchData will be broken for one day until PyTorch has updated it's nightly releases.

You find the DevEx problem that we have been suffering as torchdata is not off core.

ejguan commented on torch/utils/data/datapipes/iter/sharding.py:
```diff
@@ -0,0 +1,77 @@
+]
+
+class _ShardingIterDataPipe(IterDataPipe):
+    def is_shardable(self):
+        return True
```
TBH, I don't think this function is useful as all.

wenleix replied:
Agree. So the API contract will be: 
* `_ShardingIterDataPipe` is always shardable (after all, the name is "Sharding DataPipe"
* It has "apply_sharding" method (in the future we may introduce other things like `is_replicable`.

ejguan replied:
Yeah, exactly.

ejguan replied:
> * Adding more built-in sharding DataPipe?

IDK.Probably yes when we have a solid use case

> * Users will want to use this to implement custom sharding?

 Yeah, I think so.

wenleix replied:
Do we consider `[ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17)` as a `_ShardingIterDataPipe` ? :)

ejguan replied:
IMHO, this DataPipe should be labeled as `non-replicable` with `is_replicable` returning `False` and inherit `_ShardingIterDataPipe`.

In the `apply_sharding` function, we might want to only apply sharding to replicable `_ShardingIterDataPipe`

wenleix replied:
Makes sense to me. Perhaps a good follow-up / bootcamp task to consolidate [ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17) ?  (want to keep this PR as an initial step)

ejguan replied:
SGTM.

ejguan reviewed (approved):
LGTM. Thank you so much. We can land it from Phab

wenleix commented:
> LGTM. Thank you so much. We can land it from Phab

And TorchData needs to wait for the new PT nightly to success CI... hmm 🤔

wenleix committed:
Refactor `ShardingFilterIterDataPipe` into a seperate file (#987)

Summary:
X-link: https://github.com/pytorch/data/pull/987

Pull Request resolved: https://github.com/pytorch/pytorch/pull/94095

Differential Revision:
D43014692

Move `ShardingFilterIterDataPipe` into a dedicated file.

Also, propose to have a dedicated parent class (`_ShardingIterDataPipe`) for sharding data pipe, as this seems more like a "system/engine-level" datapipe that gives strong hints to RS on how to execute, and needs first-class citizen treatment in RS (compared with other "user-level" datapipe that are mostly composable `Callable[[Iterable], Iterable]`.  But open to other discussions.

### Open question
Should
[ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17) also be considered as a `_ShardingIterDataPipe`? (e.g. this sharding is executed by replicating (the metadata), while `ShardingRoundRobinDispatcherIterDataPipe` hints too expensive to replicate so requires round robin data exchange/dispatch).

D43014692

Test Plan:
sandcastle and CI

How to run unit tests in buck related to such changes? :)

Reviewed By: seemethere

fbshipit-source-id: fcd2a4e57b15fdf7411c959a38c377ac114f0ecb
